shared/src/ledger/native_vp.rs

```diff
@@ -0,0 +1,242 @@
+//! Native validity predicate interface associated with internal accounts such
+//! as the PoS and IBC modules.
+use std::collections::HashSet;
+
+use crate::ledger::gas::{Result, VpGasMeter};
+use crate::ledger::ibc::Ibc;
+use crate::ledger::pos::PoS;
+use crate::ledger::storage::write_log::WriteLog;
+use crate::ledger::storage::{Storage, StorageHasher};
+use crate::ledger::{storage, vp_env};
+use crate::proto::Tx;
+use crate::types::address::{Address, InternalAddress};
+use crate::types::storage::{BlockHash, BlockHeight, Key};
+use crate::vm::prefix_iter::PrefixIterators;
+
+/// Initialize genesis storage for all the [`NativeVp`]s.
+pub fn init_genesis_storage<DB, H>(storage: &mut Storage<DB, H>)
+where
+    DB: storage::DB + for<'iter> storage::DBIter<'iter>,
+    H: StorageHasher,
+{
+    PoS::init_genesis_storage(storage);
+    Ibc::init_genesis_storage(storage);
+}
+
+/// A native VP module should implement this module and add its initialization
+/// call to [`init_genesis_storage`].
+pub trait NativeVp {
+    /// The address of this VP
+    const ADDR: InternalAddress;
+
+    /// Initialize storage in the genesis block
+    fn init_genesis_storage<DB, H>(storage: &mut Storage<DB, H>)
+    where
+        DB: storage::DB + for<'iter> storage::DBIter<'iter>,
+        H: StorageHasher;
+
+    /// Run the validity predicate. This function can call methods on the
+    /// [`Ctx`] argument to interact with the host structures.
+    fn validate_tx<DB, H>(
+        ctx: &mut Ctx<DB, H>,
+        tx_data: &[u8],
+        keys_changed: &HashSet<Key>,
+        verifiers: &HashSet<Address>,
+    ) -> Result<bool>
+    where
+        DB: storage::DB + for<'iter> storage::DBIter<'iter>,
+        H: StorageHasher;
+}
+
+/// A validity predicate's host context.
+///
+/// This is similar to [`anoma_shared::vm::host_env::VpCtx`], but without the VM
+/// wrapper types and `eval_runner` field. The references must not be changed
+/// when [`Ctx`] is mutable.
+#[derive(Debug)]
+pub struct Ctx<'a, DB, H>
+where
+    DB: storage::DB + for<'iter> storage::DBIter<'iter>,
+    H: StorageHasher,
+{
+    /// Storage prefix iterators.
+    pub iterators: PrefixIterators<'a, DB>,
+    /// VP gas meter.
+    pub gas_meter: VpGasMeter,
+    /// Read-only access to the storage.
+    pub storage: &'a Storage<DB, H>,
```

Comment:
I think a native VP stores the data for the functionality to the storage.
Should we provide `write` functions?

Reply:
Thank you for reviewing! It is my understanding that any storage modifications should be done via transactions (via the write log) and then the VP checks that these are correct (same as with wasm VPs). For the native VPs integrations, we can add functions to write storage in the `vm_env` crate using the tx host env functions. These can then be called from transactions.

Reply:
Sorry, I misunderstood that. As we discussed, the VP doesn't update the storage. So, write access isn't needed.